test/replication-luatest/gh_5295_split_brain_test.lua

```diff
@@ -0,0 +1,247 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('gh-6036')
+
+local function repl_list(i)
+    local repl = {}
+    for j = 1,i do
+        table.insert(repl, server.build_instance_uri('s'..j))
+    end
+    return repl
+end
+
+-- gh-5295: the server should stop replication from an upstream which sends data
+-- conflicting in terms of promote / confirmed_lsn.
+--
+-- The test requires one instance in replicaset per each split-brain case, plus
+-- one "main" instance. The idea is to partition one instance from the main
+-- server per test case. Once partitioned, create a split-brain situation
+-- between the partitioned node and all the main's followers. Then check that
+-- the partitioned node can't reconnect.
+local num_instances = 1
```

Comment:
What if we eventually have more than 30 test case (+1 main instance, +1 is reserved for 0 vclock)? Maybe it would be more reliable and a bit easier globally to start one new instance in each case, drop it in the end, and unregister it from `_cluster`?

Reply:
Hmm. I don't know why I haven't come up with this myself.
I was too fascinated with my approach for some reason :)
Will do.

Reply:
I've reworked the test:
```diff
diff --git a/test/replication-luatest/gh_5295_split_brain_test.lua b/test/replication-luatest/gh_5295_split_brain_test.lua
index db5499016..554346489 100644
--- a/test/replication-luatest/gh_5295_split_brain_test.lua
+++ b/test/replication-luatest/gh_5295_split_brain_test.lua
@@ -4,29 +4,22 @@ local server = require('test.luatest_helpers.server')
 
 local g = t.group('gh-5295')
 
-local function repl_list(i)
-    local repl = {}
-    for j = 1, i do
-        table.insert(repl, server.build_instance_uri('s'..j))
-    end
-    return repl
-end
+-- To distinguish replicas by name
+local test_id = 0
 
 -- gh-5295: the server should stop replication from an upstream which sends data
 -- conflicting in terms of promote / confirmed_lsn.
 --
 -- The test requires one instance in replicaset per each split-brain case, plus
--- one "main" instance. The idea is to partition one instance from the main
--- server per test case. Once partitioned, create a split-brain situation
--- between the partitioned node and all the main's followers. Then check that
+-- one "main" instance. The idea of each test case is first to join the instance
+-- and then partition it from the main server. Once partitioned, create a
+-- split-brain situation between the partitioned node and main. Then check that
 -- the partitioned node can't reconnect.
-local num_instances = 1
 
 g.before_all(function(cg)
     cg.cluster = cluster:new({})
 
     cg.box_cfg = {
-        replication = repl_list(num_instances),
         replication_timeout         = 0.1,
         replication_synchro_quorum  = 1,
         replication_synchro_timeout = 0.01,
@@ -36,44 +29,63 @@ g.before_all(function(cg)
         log_level                   = 6,
     }
 
-    cg.s = {}
-    for i = 1, num_instances do
-        cg.s[i] = cg.cluster:build_and_add_server({
-            alias = 's'..i,
-            box_cfg = cg.box_cfg,
-        })
-    end
+    cg.main = cg.cluster:build_and_add_server{
+        alias = 'main',
+        box_cfg = cg.box_cfg,
+    }
     cg.cluster:start()
 
-    cg.main_id = cg.s[1]:exec(function()
+    cg.main:exec(function()
         box.ctl.promote()
         box.ctl.wait_rw()
         local s = box.schema.space.create('sync', {is_sync = true})
         s:create_index('pk')
         s = box.schema.space.create('async')
         s:create_index('pk')
-        return box.info.id
+        -- Check the test correctness.
+        assert(box.info.id == 1)
     end)
 end)
 
+local function update_replication(...)
+    return box.cfg{replication = {...}}
+end
+
 g.before_each(function(cg)
-    -- Check that the servers start synced and with s[1] being leader.
+    -- Check that the servers start synced and with main being leader.
     -- It's a prerequisite for each test.
-    cg.s[1]:exec(function()
+    cg.main:exec(function()
         require('luatest').assert_equals(box.info.synchro.queue.owner,
-                                         box.info.id, 'First node is leader')
+                                         box.info.id, 'main node is leader')
     end)
+
+    test_id = test_id + 1
+    cg.box_cfg.replication = {
+            server.build_instance_uri('main'),
+            server.build_instance_uri('split_replica'..test_id),
+    }
+    cg.split_replica = cg.cluster:build_and_add_server{
+        alias = 'split_replica'..test_id,
+        box_cfg = cg.box_cfg,
+    }
+    cg.split_replica:start()
     t.helpers.retrying({}, function()
-        cg.s[num_instances]:assert_follows_upstream(cg.main_id)
+        cg.split_replica:assert_follows_upstream(1)
+    end)
+
+    cg.main:exec(update_replication, cg.box_cfg.replication)
+    t.helpers.retrying({}, function()
+        cg.main:assert_follows_upstream(2)
     end)
 end)
 
 -- Drop the partitioned server after each case of split-brain.
 g.after_each(function(cg)
-    cg.s[num_instances]:stop()
-    cg.s[num_instances]:cleanup()
-    cg.cluster.servers[num_instances] = nil
-    num_instances = num_instances - 1
+    cg.split_replica:stop()
+    cg.split_replica:cleanup()
+    -- Drop the replica's cluster entry, so that next one receives same id.
+    cg.main:exec(function() box.space._cluster:delete{2} end)
+    cg.cluster.servers[2] = nil
 end)
 
 g.after_all(function(cg)
@@ -81,31 +93,23 @@ g.after_all(function(cg)
     cg.cluster.servers = nil
 end)
 
-local function update_replication(...)
-    return box.cfg{replication = {...}}
-end
-
-local function partition_nth_instance(cg, n)
+local function partition_replica(cg)
     -- Each partitioning starts on synced servers.
-    cg.s[n]:wait_vclock_of(cg.s[1])
-    cg.s[1]:wait_vclock_of(cg.s[n])
-    cg.s[n]:exec(update_replication, {})
-    cg.box_cfg.replication = repl_list(n - 1)
-    for i = 1, n - 1 do
-        cg.s[i]:exec(update_replication, cg.box_cfg.replication)
-    end
-    return cg.s[n]
+    cg.split_replica:wait_vclock_of(cg.main)
+    cg.main:wait_vclock_of(cg.split_replica)
+    cg.split_replica:exec(update_replication, {})
+    cg.main:exec(update_replication, {})
 end
 
-local function reconnect_and_check_split_brain(srv, main_id)
-    srv:exec(update_replication, {server.build_instance_uri('s1')})
-    t.helpers.retrying({}, srv.exec, srv, function(main_id)
-        local upstream = box.info.replication[main_id].upstream
+local function reconnect_and_check_split_brain(srv)
+    srv:exec(update_replication, {server.build_instance_uri('main')})
+    t.helpers.retrying({}, srv.exec, srv, function()
+        local upstream = box.info.replication[1].upstream
         local t = require('luatest')
         t.assert_equals(upstream.status, 'stopped', 'replication is stopped')
         t.assert_str_contains(upstream.message, 'Split-Brain discovered: ',
                               false, 'split-brain is discovered')
-    end, {main_id})
+    end)
 end
 
 local function write_promote()
@@ -130,81 +134,74 @@ local function write_demote()
     t.assert_equals(box.info.synchro.queue.owner, 0, "Demote succeeded")
 end
 
-num_instances = num_instances + 1
 -- Any async transaction performed in an obsolete term means a split-brain.
 g.test_async_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n) box.space.async:replace{n} end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function() box.space.async:replace{1} end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Any unseen sync transaction confirmation from an obsolete term means a
 -- split-brain.
 g.test_confirm_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n) box.space.sync:replace{n} end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function() box.space.sync:replace{1} end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Any unseen sync transaction rollback from an obsolete term means a
 -- split-brain.
 g.test_rollback_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function()
         box.cfg{replication_synchro_quorum = 31}
-        pcall(box.space.sync.replace, box.space.sync, {n})
+        pcall(box.space.sync.replace, box.space.sync, {1})
         box.cfg{replication_synchro_quorum = 1}
-    end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Conflicting demote for the same term is a split-brain.
 g.test_demote_same_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(write_demote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
-    cg.s[1]:exec(write_promote)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(write_demote)
+    reconnect_and_check_split_brain(cg.split_replica)
+    cg.main:exec(write_promote)
 end
 
-num_instances = num_instances + 1
 -- Conflicting promote for the same term is a split-brain.
 g.test_promote_same_term = function(cg)
-    cg.s[1]:exec(write_demote)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.main:exec(write_demote)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with lsn < confirmed_lsn is a split brain.
 g.test_promote_new_term_small_lsn = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances]) -- Make sure the test is correct.
-    test_server:exec(function(n) box.space.sync:replace{n} end, {num_instances})
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    partition_replica(cg)
+    cg.split_replica:exec(function() box.space.sync:replace{1} end)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-local function fill_queue_and_write(server, n)
-    local wal_write_count = server:exec(function(n)
+local function fill_queue_and_write(server)
+    local wal_write_count = server:exec(function()
         local fiber = require('fiber')
         box.cfg{
             replication_synchro_quorum = 31,
             replication_synchro_timeout = 1000,
         }
         local write_cnt = box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
-        fiber.new(box.space.sync.replace, box.space.sync, {n})
+        fiber.new(box.space.sync.replace, box.space.sync, {1})
         return write_cnt
-    end, {n})
+    end)
     t.helpers.retrying({}, server.exec, server, function(cnt)
         local t = require('luatest')
         local new_cnt = box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
@@ -221,27 +218,23 @@ local function perform_rollback(server)
     end)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with lsn > confirmed_lsn is a split brain.
 g.test_promote_new_term_big_lsn = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    fill_queue_and_write(cg.s[num_instances], num_instances)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances])
-    perform_rollback(test_server)
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    partition_replica(cg)
+    perform_rollback(cg.split_replica)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with conflicting queue contents is a split brain.
 g.test_promote_new_term_conflicting_queue = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    fill_queue_and_write(cg.s[num_instances], num_instances)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances])
-    perform_rollback(test_server)
-    cg.s[1]:exec(write_promote)
-    fill_queue_and_write(test_server, num_instances)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    partition_replica(cg)
+    perform_rollback(cg.split_replica)
+    cg.main:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    reconnect_and_check_split_brain(cg.split_replica)
 end

```